In [23]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import sklearn
import os
import sys
from sklearn.tree import DecisionTreeClassifier

In [24]:
dt = DecisionTreeClassifier(criterion="gini", random_state=42,
                            splitter="best",
                            max_depth=None,
                            min_samples_split=2,
                            min_samples_leaf=1,
                            min_weight_fraction_leaf=0.0,
                            max_features=None,
                            max_leaf_nodes=None,
                            min_impurity_decrease=0.0,
                            class_weight=None,
                            ccp_alpha=0.0,
                            monotonic_cst=None)

In [25]:
df = pd.DataFrame()
df_test = pd.DataFrame()
dict1 = {"LAYING":0,
"SITTING":1,
"STANDING":2,
"WALKING":3,
"WALKING_DOWNSTAIRS":4,
"WALKING_UPSTAIRS":5}

In [26]:
#Loading the UCI HAR train dataset with raw data from accelerometer
dataset_dir=fr"C:\Users\vedan\Downloads\human+activity+recognition+using+smartphones\UCI HAR Dataset\UCI HAR Dataset\Combined\Train"
for files in os.listdir(dataset_dir):
    for dfs in os.listdir(os.path.join(dataset_dir, files)):
        df1 = pd.read_csv(fr"{dataset_dir}\{files}\{dfs}")
        df1["Label"] = np.array([dict1[files]]*df1.shape[0])
        df = pd.concat([df,df1],axis = 0)

In [27]:
df.shape

(470528, 4)

In [29]:
X_train = df.drop("Label", axis=1)
Y_train = df["Label"]

In [45]:
dt.fit(X_train, Y_train)

DecisionTreeClassifier(random_state=42)

In [31]:
#Loading the UCI HAR test dataset with raw data from accelerometer
dataset_dir = fr"C:\Users\vedan\Downloads\human+activity+recognition+using+smartphones\UCI HAR Dataset\UCI HAR Dataset\Combined\Test"
for files in os.listdir(dataset_dir):
    for dfs in os.listdir(os.path.join(dataset_dir, files)):
        df1 = pd.read_csv(fr"{dataset_dir}\{files}\{dfs}")
        df1["Label"] = np.array([dict1[files]] * df1.shape[0])
        df_test = pd.concat([df_test, df1], axis=0)

In [32]:
df_test.shape

(188608, 4)

In [46]:
# Making prediction on UCI HAR test dataset
predictions = dt.predict(df_test.drop("Label", axis = 1))

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix

In [47]:

accuracy = accuracy_score(predictions, df_test["Label"])
print(f"Accuracy: {accuracy}")
precision = precision_score(predictions, df_test["Label"], average="weighted")
print(f"Precision: {precision}")
recall = recall_score(predictions, df_test["Label"], average="weighted")
print(f"Recall: {recall}")
mat = confusion_matrix(predictions, df_test["Label"])
print(f"Confusion Matrix:\n{mat}")


Accuracy: 0.5628022141160502
Precision: 0.5512171505744833
Recall: 0.5628022141160502
Confusion Matrix:
[[34278     0     0     0     0     3]
 [    0 19633  5112   350   234   232]
 [    0  5223 16564   967   482   714]
 [    1  2214  4984 12945  9046 10315]
 [   88  2183  1907  8992 10982  7133]
 [    1  2171  5481  8490  6136 11747]]


In [37]:
# Loading the collected data

# Save the current working directory
original_cwd = os.getcwd()

# Change to the directory containing the Make_collected_dataset.py file
os.chdir(os.path.abspath(r"C:\Users\vedan\Downloads\human+activity+recognition+using+smartphones\Collected Data"))

# Add the "Collected Data" directory to sys.path
sys.path.append(os.getcwd())

from Make_collected_dataset import X, Y

# Reset the working directory
os.chdir(original_cwd)


(24, 500, 3)
(24,)


In [38]:
X.shape

(24, 500, 3)

In [49]:
X_collected= X.reshape(-1, 3)

In [83]:
X_collected.shape

(12000, 3)

In [51]:
X_collected=pd.DataFrame(X_collected, columns=["accx", "accy", "accz"])

In [54]:
Y_collected=[]
for i in range(24):
    Y_collected.extend([Y[i]] * 500)
Y_collected=np.array(Y_collected)

In [55]:
Y_collected.shape

(12000,)

In [52]:
# Making prediction on collected dataset using tree trained on UCI HAR dataset
predictions2 = dt.predict(X_collected)

In [56]:

accuracy = accuracy_score(predictions2, Y_collected)
print(f"Accuracy: {accuracy}")
precision = precision_score(predictions2, Y_collected, average="weighted")
print(f"Precision: {precision}")
recall = recall_score(predictions2, Y_collected, average="weighted")
print(f"Recall: {recall}")

mat2 = confusion_matrix(predictions2, Y_collected)
print(f"Confusion Matrix:\n{mat2}")

Accuracy: 0.16991666666666666
Precision: 0.48228979166666675
Recall: 0.16991666666666666
Confusion Matrix:
[[   0  941  765 1033   85  332   71]
 [   0  112  181   99    0    2    0]
 [   0    0    0    0    0    0    0]
 [   0  175  248  157  170  164  211]
 [   0  545  545  503 1739 1471 1713]
 [   0  227  261  208    6   31    5]
 [   0    0    0    0    0    0    0]]


C:\Users\vedan\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\vedan\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [57]:
# Loading TSFEL featurised data of UCI HAR Dataset
Xtrain_tsfel = pd.read_csv(fr"UCI HAR Dataset\X_train_tsfel.csv")
Ytrain_tsfel = pd.read_csv(fr"UCI HAR Dataset\y_train_tsfel.csv",header = None)
Xtest_tsfel = pd.read_csv(fr"UCI HAR Dataset\X_test_tsfel.csv")
Ytest_tsfel = pd.read_csv(fr"UCI HAR Dataset\y_test_tsfel.csv",header= None)

In [65]:
#Making prediction on TSFEL featurised test data of UCI HAR Dataset
dt.fit(Xtrain_tsfel, Ytrain_tsfel)
predictions3 = dt.predict(Xtest_tsfel)

In [64]:
accuracy = accuracy_score(predictions3, Ytest_tsfel)
print(f"Accuracy: {accuracy}")

precision = precision_score(predictions3, Ytest_tsfel,average="weighted")
print(f"Precision: {precision}")

recall = recall_score(predictions3, Ytest_tsfel,average="weighted")
print(f"Recall: {recall}")

mat3 = confusion_matrix(predictions3, Ytest_tsfel)
print(f"Confusion Matrix:\n{mat3}")

Accuracy: 0.8888888888888888
Precision: 0.897119341563786
Recall: 0.8888888888888888
Confusion Matrix:
[[6 1 0 0 0 0]
 [1 8 2 0 0 0]
 [2 0 7 0 0 0]
 [0 0 0 9 0 0]
 [0 0 0 0 9 0]
 [0 0 0 0 0 9]]


In [61]:
# Loading TSFEL featurised data of Collected Dataset
Xcollected_tsfel = pd.read_csv(fr"Collected Data\X_collected_tsfel.csv")
Ycollected_tsfel = pd.read_csv(fr"Collected Data\y_collected_tsfel.csv",header = None)

In [66]:
#Making prediction on TSFEL featurised collected data using tree trained on tsfel featurised UCI HAR dataset 
predictions4 = dt.predict(Xcollected_tsfel)

In [67]:
accuracy = accuracy_score(predictions4, Ycollected_tsfel)
print(f"Accuracy: {accuracy}")

precision = precision_score(predictions4, Ycollected_tsfel, average="weighted")
print(f"Precision: {precision}")

recall = recall_score(predictions4, Ycollected_tsfel, average="weighted")
print(f"Recall: {recall}")

mat4 = confusion_matrix(predictions4, Ycollected_tsfel)
print(f"Confusion Matrix:\n{mat4}")

Accuracy: 0.20833333333333334
Precision: 0.8020833333333334
Recall: 0.20833333333333334
Confusion Matrix:
[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 1 0 0]
 [2 1 1 0 0 0]
 [2 3 3 3 4 4]]


C:\Users\vedan\AppData\Roaming\Python\Python312\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


The Model was first trained on raw accelerometer data of UCI HAR Dataset and used to make predictions on the collected data where it got 17% accuracy with a precision of 48% and recall 17%.
Then it was trained on TSFEL featurised UCI HAR Dataset and used to make predictions on the TSFEL featurised collected data where it got 20% accuracy with a precision of 80% and recall 20%.
Possible reasons for the accuracy scores could be different orientation of the accelerometer in the UCI experiment and during our data collection, accelerometer senstivities, preprocessing used. 